pkg/ingester/client/circuitbreaker.go

```diff
@@ -0,0 +1,72 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package client
+
+import (
+	"context"
+	"errors"
+
+	"github.com/sony/gobreaker"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	resultSuccess = "success"
+	resultError   = "error"
+	resultOpen    = "circuit_breaker_open"
+)
+
+func NewCircuitBreaker(name string, cfg CircuitBreakerConfig, metrics *Metrics) grpc.UnaryClientInterceptor {
+	breaker := gobreaker.NewCircuitBreaker(gobreaker.Settings{
+		Name:        name,
+		Interval:    cfg.ClosedInterval,
+		Timeout:     cfg.OpenTimeout,
+		MaxRequests: uint32(cfg.MaxHalfOpenRequests),
+		ReadyToTrip: func(counts gobreaker.Counts) bool {
+			return uint64(counts.ConsecutiveFailures) >= cfg.MaxConsecutiveFailures
+		},
+		OnStateChange: func(name string, from gobreaker.State, to gobreaker.State) {
+			metrics.circuitBreakerTransitions.WithLabelValues(to.String()).Inc()
+		},
+		IsSuccessful: isSuccessful,
+	})
+
+	// Initialize each of the known labels for circuit breaker metrics
+	metrics.circuitBreakerTransitions.WithLabelValues(gobreaker.StateOpen.String())
+	metrics.circuitBreakerTransitions.WithLabelValues(gobreaker.StateHalfOpen.String())
+	metrics.circuitBreakerTransitions.WithLabelValues(gobreaker.StateClosed.String())
+	metrics.circuitBreakerResults.WithLabelValues(resultSuccess)
+	metrics.circuitBreakerResults.WithLabelValues(resultError)
+	metrics.circuitBreakerResults.WithLabelValues(resultOpen)
+
+	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		_, err := breaker.Execute(func() (interface{}, error) {
+			err := invoker(ctx, method, req, reply, cc, opts...)
+			return nil, err
+		})
+
+		if err != nil && (errors.Is(err, gobreaker.ErrOpenState) || errors.Is(err, gobreaker.ErrTooManyRequests)) {
+			metrics.circuitBreakerResults.WithLabelValues(resultOpen).Inc()
+		} else if !isSuccessful(err) {
+			metrics.circuitBreakerResults.WithLabelValues(resultError).Inc()
+		} else {
+			metrics.circuitBreakerResults.WithLabelValues(resultSuccess).Inc()
+		}
+
+		return err
```

Comment:
Not sure if the error returned here should get the `globalerror` treatment with a helpful message and everything.

Reply:
I don't know neither, but I would start easy (no `globalerror` treatment): it's something we can revisit anytime.